core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -0,0 +1,267 @@
+      private case object Poll
+      // FIXME move out settings
+      private val POLL_INTERVAL = 100.millis
+      private val POLL_DATA_TIMEOUT = 50.millis
+      private val POLL_COMMIT_TIMEOUT = 1.millis
+      private val COMMIT_TIMEOUT = 10.seconds
+      private var awaitingConfirmation = 0L
+      private var buffer: Iterator[Consumer.CommittableMessage[K, V]] = Iterator.empty
+      private var pollScheduled = false
+
+      private def schedulePoll() = {
+        if (!pollScheduled) {
+          pollScheduled = true
+          scheduleOnce(Poll, POLL_INTERVAL)
+        }
+      }
+
+      private def poll() = {
+        def setupConsumer() = {
+          // TODO do we really need to support change of assignments after startup?
+          //      Lots of allocations here.
+          if (isAvailable(out)) {
+            consumer.resume(consumer.assignment().asScala.toSeq: _*)
+          }
+          else {
+            consumer.pause(consumer.assignment().asScala.toSeq: _*)
+          }
+        }
+
+        def handleResult(records: ConsumerRecords[K, V]) = {
+          if (!records.isEmpty) {
+            logger.trace(s"Got messages - {}", records)
+            require(!buffer.hasNext)
+            require(isAvailable(out))
+            buffer = records.iterator().asScala.map { record =>
+              val offset = Consumer.Offset(
+                topic = record.topic,
+                partition = record.partition,
+                offset = record.offset
+              )
+              val committable = CommittableOffsetImpl(offset)(this)
+              Consumer.CommittableMessage(record.key, record.value, committable)
+            }
+            pushMsg(buffer.next())
+          }
+        }
+
+        def pollTimeout = {
+          val data = Option(POLL_DATA_TIMEOUT).filter(_ => isAvailable(out) && !buffer.hasNext)
+          def commit = Option(POLL_COMMIT_TIMEOUT).filter(_ => awaitingConfirmation > 0)
+          data.orElse(commit)
+        }
+
+        pollTimeout match {
+          case Some(timeout) =>
+            setupConsumer()
+            // TODO poll is blocking, can't we do this without blocking?
+            val msgs = consumer.poll(timeout.toMillis)
+            handleResult(msgs)
+            if (pollTimeout.isDefined) schedulePoll()
+          case None =>
+        }
+
+        if (stopping && !isClosed(out)) {
+          logger.debug("Stop producing messages")
+          complete(out)
+        }
+
+        if (stopping && awaitingConfirmation == 0) {
+          completeStage()
+        }
+      }
+
+      val pollCallback = getAsyncCallback[Unit] { _ => schedulePoll() }
+
+      private def pushMsg(msg: Consumer.CommittableMessage[K, V]): Unit = {
+        logger.trace("Push element {}", msg)
+        push(out, msg)
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          if (!buffer.hasNext) poll()
```

Comment:
I agree, it can improved. But we also should pay attention to how it optimized in kafka client. There is also prefetch algorithm in it. Check sources.